### PR TITLE
Add DestinationConsumer->SQS, Sequin.Aws

### PR DESCRIPTION
**Pull request by acco**

### CODE REVIEW

**Comment by acco** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/542?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#542** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/542?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#541** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/541?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#540** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/540?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#539** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/539?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @acco and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lib/sequin/aws/queue_policy.ex [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        account_id: account_id,
+        region: region,
+        resource: "arn:aws:sqs:#{region}:#{account_id}:#{queue_name}"
+      },
+      fn {account_id, _} ->
+        %{"Effect" => "Allow", "Principal" => %{"AWS" => "arn:aws:iam::#{account_id}:root"}, "Action" => "SQS:*"}
+      end
+    )
+
+    # By default, let the root act on the queue
```
Comment (RTLS) [inline lib/sequin/aws/queue_policy.ex:50]: ?

File: lib/sequin/aws/sqs.ex [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+           receipt_handle: String.t(),
+           message_body: String.t()
+         }
+
+  @doc """
+  Call with the desired queue name and QueuePolicy. Returns {:ok, %{arn: queue_arn, url: queue_url}}.
+  """
+  @spec create_queue(Client.t(), String.t(), QueuePolicy.t(), keyword()) ::
+          {:ok, map()} | {:error, any()}
+  def create_queue(client, queue_name, %QueuePolicy{} = queue_policy, opts) do
```
Comment (RTLS) [inline lib/sequin/aws/sqs.ex:20]: based on this we are doing queue management? or is this just for dev/test?

File: lib/sequin/aws/sqs.ex [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+              "Name" => to_string(k),
+              "Value.StringValue" => to_string(v),
+              "Value.DataType" => "String"
+            }
+          end)
+
+        attributes = to_aws_object_array("MessageAttribute", attributes)
+
+        %{
+          "Id" => msg[:id] || UUID.uuid4(),
```
Comment (RTLS) [inline lib/sequin/aws/sqs.ex:92]: when do we want to fallback to a generated uuid?

File: lib/sequin/aws/sqs.ex [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+      |> Map.put("QueueUrl", queue_url)
+
+    case AWS.SQS.send_message_batch(client, wrapped_messages) do
+      {:ok,
+       %{
+         "SendMessageBatchResponse" => %{
+           "SendMessageBatchResult" => %{"BatchResultErrorEntry" => _}
+         }
+       } = resp, %{body: body, status_code: 200}} ->
+        {:error, resp, %{body: body}}
```
Comment (RTLS) [inline lib/sequin/aws/sqs.ex:112]: cool a 200 with error in the body 👍

File: lib/sequin/aws/sqs.ex [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+        {:ok, messages}
+
+      {:error,
+       {:unexpected_response,
+        %{
+          body: body,
+          status_code: 400
+        }}} = err ->
+        if String.contains?(body, "AWS.SimpleQueueService.NonExistentQueue") do
+          {:error, :non_existent_queue}
```
Comment (RTLS) [inline lib/sequin/aws/sqs.ex:153]: Error.bad_request or not_found ?

File: lib/sequin/consumers/sqs_destination.ex [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  defp put_is_fifo(changeset) do
+    if changeset |> get_field(:queue_url) |> ends_with_fifo?() do
+      put_change(changeset, :is_fifo, true)
+    else
+      changeset
+    end
+  end
+
+  defp ends_with_fifo?(nil), do: false
+  defp ends_with_fifo?(url), do: String.ends_with?(url, ".fifo")
```
Comment (RTLS) [inline lib/sequin/consumers/sqs_destination.ex:45]: cool

File: lib/sequin/consumers/sqs_destination.ex [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    struct
+    |> cast(params, [:queue_url, :region, :endpoint, :access_key_id, :secret_access_key, :is_fifo])
+    |> validate_required([:queue_url, :region, :access_key_id, :secret_access_key])
+    |> validate_queue_url()
+    |> put_is_fifo()
+  end
+
+  defp validate_queue_url(changeset) do
+    changeset
+    |> validate_format(:queue_url, ~r/^https?:\/\/.+/, message: "must be a valid URL")
```
Comment (RTLS) [inline lib/sequin/consumers/sqs_destination.ex:32]: should these always be aws urls? or they can vary with custom dns routing probably

File: test/sequin/aws_sqs_test.exs [hunk 8]
```diff
@@ -0,0 +1,93 @@
+defmodule Sequin.Aws.SQSTest do
+  use ExUnit.Case, async: true
+
+  alias Sequin.Aws.HttpClient
+  alias Sequin.Aws.SQS
+  alias Sequin.Factory.Destination
```
Comment (RTLS) [inline test/sequin/aws_sqs_test.exs:6]: `DestinationFactory` ?

**Pull request merged**